### PR TITLE
Fix a few AMDGPU related issues

### PR DESCRIPTION
**Pull request by DiamondLovesYou**
* AMDGPU ignores `noinline` and sadly doesn't clear the attribute when it slaps `alwaysinline` on everything,
* an AMDGPU related load bit range metadata assertion,
* I didn't enable the `amdgpu` component in the `librustc_llvm` build script,
* Add AMDGPU call abi info.

### CODE REVIEW

**Comment by pietroalbini** [comment]
Ping from triage @eddyb! This PR needs your review.

**Comment by pietroalbini** [comment]
Ping from triage! This PR needs a review, can @eddyb or someone else from @rust-lang/compiler review this?

**Comment by estebank** [comment]
The code itself looks good to me. Leaving to somebody else that has better knowledge of llvm to chime in and r+.

File: src/librustc_target/spec/mod.rs [hunk 1]
```diff
@@ -635,6 +635,12 @@ pub struct TargetOptions {
     /// typically because the platform needs to unwind for things like stack
     /// unwinders.
     pub requires_uwtable: bool,
+
+    /// Targets like AMDGPU add alwaysinline hints everywhere,
+    /// and basically inline everything regardless of our wishes anyway.
+    /// This causes LLVM validation errors due to the conflicting attributes.
+    /// Defaults to false.
+    pub ignore_inline_never: bool,
```
Comment (nagisa) [inline src/librustc_target/spec/mod.rs]: General strategy is to have `is_like_platform` for platform-specific handling.

Comment (nagisa) [reply src/librustc_target/spec/mod.rs]: This seems fine to me, though.

File: src/librustc_codegen_llvm/builder.rs [hunk 2]
```diff
@@ -564,6 +564,14 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
 
     pub fn range_metadata(&self, load: ValueRef, range: Range<u128>) {
+        if self.sess().target.target.arch == "amdgpu" {
+            // amdgpu/LLVM does something weird and thinks a i64 value is
```
Comment (nagisa) [inline src/librustc_codegen_llvm/builder.rs]: This sounds like a bug in LLVM. Consider filling a (rust-lang) issue and referencing this piece of code there.

It is definitely desirable to maintain range metadata somehow.

File: src/librustc_target/spec/mod.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_target/spec/mod.rs]: Please don't do this, if it's a LLVM-specific *temporary hack*, add code for it checking e.g. architecture name in triple, to `rustc_codegen_llvm`.

File: src/librustc_codegen_llvm/builder.rs [hunk 3]
```diff
@@ -564,6 +564,14 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
 
     pub fn range_metadata(&self, load: ValueRef, range: Range<u128>) {
+        if self.sess().target.target.arch == "amdgpu" {
```
Comment (eddyb) [inline src/librustc_codegen_llvm/builder.rs]: Basically use a check like this instead of `ignore_inline_never`. Both are LLVM bugs AFAICT.

**Comment by Aaronepower** [comment]
Triage: @DiamondLovesYou We haven't heard from you, Are you able to get back to this?

**Comment by DiamondLovesYou** [comment]
Sorry, I've been out of the States for a while. I've returned but won't be all the way back home till Sunday.

**Comment by Aaronepower** [comment]
Triage: @DiamondLovesYou Hope your trip went well. Do you have any status updates on this PR?

**Comment by DiamondLovesYou** [comment]
Ready for the second round of reviews. @eddyb @estebank

**Comment by pietroalbini** [comment]
Ping from triage @eddyb! This PR needs your review.

**Comment by TimNN** [comment]
Ping from triage @eddyb / @estebank / @rust-lang/compiler: This PR requires your review.

**Comment by estebank** [comment]
LGTM but deferring to @eddyb.

**Comment by TimNN** [comment]
Ping from triage @eddyb / @rust-lang/compiler: This PR requires your review.

**Comment by eddyb** [comment]
Sorry about the delay, was caught up in edition-related changes.

@bors r+

**Pull request merged**